File: ICPC/WarmUp 10/ncpc2015solutions/justaquiz_lukas.cs
//Solution by lukasP (Lukáš Poláček)
using System;
using System.IO;
using System.Collections.Generic;
using System.Linq;
using System.Globalization;

class Problem
{
    class Node {
        public Dictionary<string, int> next = new Dictionary<string, int> ();
        public Dictionary<string, int> answers = new Dictionary<string, int> ();
        public double probability { get; set; }  // probability of being right if picking the most common answer in the subtree
        public int total_answers { get; set; }
    }

    static List<Node> nodes = Enumerable.Repeat(new Node(), 1).ToList();
    static double Traverse(int j, double[] best, int bpos, int left)
    {
        if (left == 0) return 0;
        double now = nodes[j].probability + best[bpos]; // answer now and move to the next question
        double move = nodes[j].next.Values
            .Sum(x => Traverse(x, best, bpos + 1, left - 1) * nodes[x].total_answers);

        return Math.Max(move / nodes[j].total_answers, now);
    }

    static void AddToNode(string answer, int index)
    {
        if (nodes[index].answers.ContainsKey(answer)) nodes[index].answers[answer]++;
        else nodes[index].answers.Add(answer, 1);
    }

    public static void Main ()
    {
        var bs = new BufferedStream(Console.OpenStandardInput());
        StreamReader reader = new StreamReader(bs);
        int[] y = reader.ReadLine().Split().Select(int.Parse).ToArray();
        int t = y[0], n = y[1];

        foreach (int i in Enumerable.Range(0, n))
        {
            var words = reader.ReadLine().Split();
            string answer = words.Last();

            int index = 0;
            AddToNode(answer, 0);
            foreach (string cur in words.Take(words.Count() - 1))
            {
                if (!nodes[index].next.ContainsKey(cur))
                {
                    nodes[index].next.Add(cur, nodes.Count());
                    nodes.Add(new Node());
                }
                index = nodes[index].next[cur];
                AddToNode(answer, index);
            }
        }

        foreach (Node node in nodes)
        {
            int s = node.answers.Values.Sum();
            node.probability = node.answers.Values.Max() / (double)s;
            node.total_answers = s;
        }

        double[] best = new double[t + 1];
        for (int i = t - 1; i >= 0; i--)
            best[i] = Traverse(0, best, i + 1, t - i);

        Console.WriteLine(best[0].ToString("#.############", CultureInfo.GetCultureInfo("en-GB")));
    }
}

File: ICPC/WarmUp 10/ncpc2015solutions/justaquiz_lukas.cpp
//Solution by lukasP (Lukáš Poláček)
#include <iostream>
#include <sstream>
#include <vector>
#include <string>
#include <unordered_map>
#include <cstdio>
using namespace std;

#define rep(i,a,b) for(__typeof(b) i=a; i<(b); ++i)

struct Node {
    unordered_map<string, int> next;
    unordered_map<string, int> answers;
    double probability; // probability of being right if picking the most common answer in the subtree
    int total_answers;
};
vector<Node> nodes(1, Node());
double traverse(int j, vector<double>::iterator it, int left)
{
    if (left == 0) return 0;
    double now = nodes[j].probability + *it; // answer now and move to the next question
    double move = 0;
    for (auto& w : nodes[j].next) // wait for the next word
        move += traverse(w.second, it + 1, left - 1) * nodes[w.second].total_answers;

    return max(now, move / nodes[j].total_answers);
}

int main()
{
    int t, n; scanf("%d %d", &t, &n);
    string s; getline(cin, s);
    rep(i,0,n)
    {
        getline(cin, s);
        stringstream w(s);
        vector<string> words;
        while (w >> s) words.push_back(s);

        string answer = words.back();
        words.pop_back();
        nodes[0].answers[answer]++;

        int index = 0;
        for (string &cur : words)
        {
            if (!nodes[index].next.count(cur))
            {
                nodes[index].next[cur] = nodes.size();
                nodes.push_back(Node());
            }
            index = nodes[index].next[cur];
            nodes[index].answers[answer]++;
        }
    }

    for (Node &node : nodes)
    {
        int s = 0, ma = 0;
        for (auto& w : node.answers)
        {
            s += w.second;
            ma = max(ma, w.second);
        }
        node.probability = ma / double(s);
        node.total_answers = s;
    }

    vector<double> best(t + 1);
    for (int i = t - 1; i >= 0; i--)
        best[i] = traverse(0, best.begin() + i + 1, t - i);

    printf("%.10lf\n", best[0]);
}
